internal/resources/kustomization/resource_kustomization.go

```diff
@@ -40,3 +55,427 @@ var kustomizationSchema = map[string]*schema.Schema{
+
+	kustomizationName, ok := d.Get(nameKey).(string)
+	if !ok {
+		return diag.Errorf("unable to read kustomization name")
+	}
+
+	kustomizationNamespaceName, ok := d.Get(namespaceNameKey).(string)
+	if !ok {
+		return diag.Errorf("unable to read kustomization namespace name")
+	}
+
+	scopedFullnameData := scope.ConstructScope(d, kustomizationName, kustomizationNamespaceName)
+
+	if scopedFullnameData == nil {
+		return diag.Errorf("Unable to get Tanzu Mission Control kustomization entry; Scope full name is empty")
+	}
+
+	UID, meta, atomicSpec, clusterScopeStatus, clusterGroupScopeStatus, err := retrieveKustomizationUIDMetaAndSpecFromServer(config, scopedFullnameData, d)
+	if err != nil {
+		if clienterrors.IsNotFoundError(err) {
+			_ = schema.RemoveFromState(d, m)
+			return
+		}
+
+		return diag.FromErr(err)
+	}
+
+	// always run
+	d.SetId(UID)
+
+	if err := d.Set(common.MetaKey, common.FlattenMeta(meta)); err != nil {
+		return diag.FromErr(err)
+	}
+
+	var (
+		flattenedSpec   []interface{}
+		flattenedStatus interface{}
+	)
+
+	switch scopedFullnameData.Scope {
+	case commonscope.ClusterScope:
+		flattenedSpec = spec.FlattenSpecForClusterScope(atomicSpec)
+		flattenedStatus = status.FlattenStatusForClusterScope(clusterScopeStatus)
+	case commonscope.ClusterGroupScope:
+		clusterGroupScopeSpec := &kustomizationclustergroupmodel.VmwareTanzuManageV1alpha1ClustergroupNamespaceFluxcdKustomizationSpec{
+			AtomicSpec: atomicSpec,
+		}
+		flattenedSpec = spec.FlattenSpecForClusterGroupScope(clusterGroupScopeSpec)
+		flattenedStatus = status.FlattenStatusForClusterGroupScope(clusterGroupScopeStatus)
+	}
+
+	if err := d.Set(spec.SpecKey, flattenedSpec); err != nil {
+		return diag.FromErr(err)
+	}
+
+	if err := d.Set(statusKey, flattenedStatus); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return diags
+}
+
+// nolint: gocognit
+func retrieveKustomizationUIDMetaAndSpecFromServer(config authctx.TanzuContext, scopedFullnameData *scope.ScopedFullname, d *schema.ResourceData) (
+	string,
+	*objectmetamodel.VmwareTanzuCoreV1alpha1ObjectMeta,
+	*kustomizationclustermodel.VmwareTanzuManageV1alpha1ClusterNamespaceFluxcdKustomizationSpec,
+	*kustomizationclustermodel.VmwareTanzuManageV1alpha1ClusterNamespaceFluxcdKustomizationStatus,
+	*kustomizationclustergroupmodel.VmwareTanzuManageV1alpha1ClustergroupNamespaceFluxcdKustomizationStatus,
+	error) {
+	var (
+		UID                     string
+		meta                    *objectmetamodel.VmwareTanzuCoreV1alpha1ObjectMeta
+		spec                    *kustomizationclustermodel.VmwareTanzuManageV1alpha1ClusterNamespaceFluxcdKustomizationSpec
+		clusterScopeStatus      *kustomizationclustermodel.VmwareTanzuManageV1alpha1ClusterNamespaceFluxcdKustomizationStatus
+		clusterGroupScopeStatus *kustomizationclustergroupmodel.VmwareTanzuManageV1alpha1ClustergroupNamespaceFluxcdKustomizationStatus
+	)
+	// nolint: dupl
+	switch scopedFullnameData.Scope {
+	case commonscope.ClusterScope:
```

Comment:
code duplication can be avoided here

Reply:
Updated.

Reply:
how are you avoiding duplicates ?

Reply:
I have moved the common logic out of the switch case right after it.